File: app/models/research.py
"""文献检索下载 Agent 的 SQLAlchemy 持久化模型。"""

from datetime import datetime, timezone

from sqlalchemy import Column, DateTime, ForeignKey, Integer, String, Text

from app.database.base import Base


def utcnow() -> datetime:
    return datetime.now(timezone.utc)


class PaperImport(Base):
    """一次「搜索到下载入库」任务。"""

    __tablename__ = "paper_imports"

    id = Column(Integer, primary_key=True, autoincrement=True)
    title = Column(String(512), nullable=False, default="")
    year = Column(Integer, nullable=True)  # 发表年份(前端从 SearchResult.year 带入, L1.5 free_pdf 按年定位会议/卷页用)
    venue = Column(String(256), nullable=False, default="")  # 发表 venue(前端带入, L1.5 free_pdf 按 venue 路由免费源)
    source = Column(String(32), nullable=False, default="")  # arxiv | semantic_scholar
    external_id = Column(String(128), nullable=True)  # arxiv_id / S2 paperId
    doi = Column(String(256), nullable=True)
    pdf_url = Column(String(1024), nullable=True)
    page_url = Column(String(1024), nullable=True)
    status = Column(String(32), nullable=False, default="pending", index=True)
    progress = Column(Integer, nullable=False, default=0)
    error_message = Column(Text, nullable=False, default="")
    paper_id = Column(Integer, ForeignKey("papers.id", ondelete="SET NULL"), nullable=True, index=True)  # 入库成功后关联 papers.id
    created_at = Column(DateTime, nullable=False, default=utcnow)
    updated_at = Column(DateTime, nullable=False, default=utcnow, onupdate=utcnow)

File: web/src/types/research.ts
/** 文献检索 Agent 类型定义 */

export type SearchSource = "arxiv" | "semantic_scholar" | "openalex";
export type OaStatus = "open" | "closed" | "unknown";

export interface SearchResult {
  source: SearchSource;
  title: string;
  authors: string[];
  year: number | null;
  venue: string;
  abstract: string;
  doi: string | null;
  pdf_url: string | null;
  page_url: string;
  citations: number;
  published: boolean;
  ccf_level: "A" | "B" | "C" | null;
  oa_status: OaStatus;
  openalex_id: string | null;
}

export type ImportStatus = "pending" | "downloading" | "parsing" | "done" | "failed";

export interface ImportTask {
  id: number;
  title: string;
  source: string;
  status: ImportStatus;
  progress: number;
  error_message: string;
  paper_id: number | null;
  created_at: string;
  updated_at: string;
}

export type BrowserStatusType = "none" | "alive" | "expired";

export interface BrowserStatus {
  status: BrowserStatusType;
  message: string;
}

export interface ResearchPlanEvent {
  event: "plan";
  queries: string[];
  sources: string[];
  direct?: boolean;
}

export interface ResearchResultsEvent {
  event: "results";
  items: SearchResult[];
  total: number;
  offset: number;
  total_is_estimate?: boolean;
}

export interface ResearchErrorEvent {
  event: "error";
  message: string;
}

export type ResearchProgressEvent =
  | ResearchPlanEvent
  | ResearchResultsEvent
  | ResearchErrorEvent
  | { event: "done" };
